tests/riscv64/deeplabv3_float_rvv_test.run

```diff
@@ -0,0 +1,3 @@
+// RUN: ${TEST_MODULE_CMD} --module_file="${BUILD_RISCV_DIR}/deeplabv3.tflite.rvv.vmfb" \
+// RUN:   --entry_function=main --function_input="1x257x257x3xf32=0"
+// iree-run-module can't printout the output selectively. Skip the numerical check.
```

Comment:
What test coverage are you actually looking to add with this? We should prefer to add unit tests, not integration tests. We already have a fairly comprehensive test suite running on most platforms, and it doesn't make sense to add platform specific integration tests for entire programs unless there are specific issues with those programs on those platforms that we can't easily reduce down to smaller tests. It could make sense to add integration tests for entire programs that run on all platforms though.

Reply:
This is the request from @dcaballe in https://github.com/iree-org/iree/issues/9824. My understanding is that these models may have wrong inference result when we make upstream llvm/mlir changes or bump the llvm-project version. The tests here are aimed to catch these breakages sooner.

@dcaballe Do you care to elaborate?

Reply:
Yes and I objected when that test was added also :-) https://github.com/iree-org/iree/pull/8304. This also goes from one test to many.

IREE e2e model tests are not the right place to be testing the LLVM repository. If there's missing test coverage there, that should be addressed upstream.

Reply:
Generally speaking, we should be aiming for a "pyramid" of tests within each project (searching for that term turns up a bunch of blog posts). Most of our coverage should come from unit tests exercising individual, tiny features. Unit tests provide the most direct error messages and are easy to iterate on and author, but they can miss interactions between features. Integration tests are a step up - they use multiple features together. At the tip of the pyramid are e2e tests, UI tests, manual tests, etc. - these can test the whole system, access external resources, step through user interface flows, etc. but are difficult to write and difficult to maintain.

In the context of IREE,

* unit tests are compiler lit tests using `iree-opt`/`iree-compile` and runtime GoogleTest (`cc_test`) tests
* integration tests are roughly what's in [`tests/e2e/`](https://github.com/iree-org/iree/tree/main/tests/e2e) using `iree_check_test`
* the third category would be things like benchmarks or comparisons between frameworks (including Python tests that use Jax/TF/PyTorch and IREE)

The best ways to increase test coverage on new platforms are to run the existing unit and integration test suites on those platforms. I like to think of bringing up a new service/platform/system as going first from "0->1", then "1->2", then "2->N". The RISC-V testing in IREE is at "2" now, and this PR tries to go from "2->4". We should skip over "4, 5, 6, 7, 8, etc." and just go straight to "N" :)

If there's value in testing numerical correctness for these specific programs, we should add that coverage for all platforms in some way, then we would get it "for free" on RISC-V once RISC-V runs the same set of tests as the rest of the project. We don't want to be fixing regressions in a specific program for a specific platform without knowing if those regressions also occur on other platforms.

Reply:
> If there's missing test coverage there, that should be addressed upstream.

I totally agree with this but we need a way to catch those issues and unfortunately some of those issues are the result of the interaction between IREE and LLVM. Not doing so means that we would be inadvertently integrating bugs that we would have to later root caused from scratch. These bugs may not only lead to correctness issues but also impact performance that we are reporting to our partners. Whatever the implementation strategy is, I think we should play a more defensive role here.

I also agree with what @ScottTodd mentioned. I think we all agree that adding models as e2e test shouldn't be a replacement for unit tests or a source for debugging. However, e2e models will help us strengthen out unit test coverage by catching issues that couldn't be caught otherwise. I'm happy with whatever approach we take as long as we can get more e2e correctness coverage. Adding them in a way so that other platforms can get them "for free' makes all the sense to me. However, it looks like the current approach requires that the models are modified and the reference output is hard-coded somehow. It's not clear how we could extend this to make it more flexible. Would it be a way to specify a reference backend implementation to compare against (with a pre-determined error tolerance)?

Thanks!

Reply:
> Generally speaking, we should be aiming for a "pyramid" of tests within each project (searching for that term turns up a bunch of blog posts). Most of our coverage should come from unit tests exercising individual, tiny features. Unit tests provide the most direct error messages and are easy to iterate on and author, but they can miss interactions between features. Integration tests are a step up - they use multiple features together. At the tip of the pyramid are e2e tests, UI tests, manual tests, etc. - these can test the whole system, access external resources, step through user interface flows, etc. but are difficult to write and difficult to maintain.
> 
> In the context of IREE,
> 
> * unit tests are compiler lit tests using `iree-opt`/`iree-compile` and runtime GoogleTest (`cc_test`) tests
> * integration tests are roughly what's in [`tests/e2e/`](https://github.com/iree-org/iree/tree/main/tests/e2e) using `iree_check_test`
> * the third category would be things like benchmarks or comparisons between frameworks (including Python tests that use Jax/TF/PyTorch and IREE)
> 
> The best ways to increase test coverage on new platforms are to run the existing unit and integration test suites on those platforms. I like to think of bringing up a new service/platform/system as going first from "0->1", then "1->2", then "2->N". The RISC-V testing in IREE is at "2" now, and this PR tries to go from "2->4". We should skip over "4, 5, 6, 7, 8, etc." and just go straight to "N" :)
> 
> If there's value in testing numerical correctness for these specific programs, we should add that coverage for all platforms in some way, then we would get it "for free" on RISC-V once RISC-V runs the same set of tests as the rest of the project. We don't want to be fixing regressions in a specific program for a specific platform without knowing if those regressions also occur on other platforms.

https://github.com/iree-org/iree/pull/10330 should bridge the gap at "1". Regardless the conclusion of the models here, I think we can increase the RISCV test coverage with that PR.